### PR TITLE
Backport: Block template utils and rest templates controller

### PR DESCRIPTION
**Pull request by ntsekouras**
Trac ticket: https://core.trac.wordpress.org/ticket/56467

Backports changes from: 
-  https://github.com/WordPress/gutenberg/pull/41387
- https://github.com/WordPress/gutenberg/pull/41189
- https://github.com/WordPress/gutenberg/pull/42520
- https://github.com/WordPress/gutenberg/pull/42586

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/rest-api/wpRestTemplatesController.php
Comment (ntsekouras) [inline tests/phpunit/tests/rest-api/wpRestTemplatesController.php]: How do you suggest to make these tests for core? In GB we have a different approach: https://github.com/WordPress/gutenberg/pull/42520/files#diff-5dcd7f0ba0e053098e93f6907913da73787fb7487e05d37e3249fb1c0ec434d6.

Should we create a new test theme or update an existing one or something else? 🤔 --cc @spacedmonkey

Comment (ntsekouras) [reply tests/phpunit/tests/rest-api/wpRestTemplatesController.php]: Used 'block-theme' eventually and tested with `slugs`.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Is there a unit test for this?

Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Added here: https://github.com/WordPress/wordpress-develop/pull/3221/commits/98878136a99e9c0e55376209f83833d9050e5e9e

**Comment by ntsekouras** [comment]
~Do not merge before this GB PR: https://github.com/WordPress/gutenberg/pull/44085/~

---
edit: 

The GB PR has merged.

**Review by ockham — COMMENTED** [review]
Thank you, @ntsekouras! This LGTM; however, we need to wait for Gutenberg 14.1 (which is the first release to contain WordPress/gutenberg#44085) to be released before merging this PR. Otherwise, a WP install running Core `trunk` and Gutenberg stable will fatal because of the duplicate `get_template_hierarchy` definition.

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php:1047]: Shouldn't this be prefixed with `wp_`? Why? 

- consistency
- to avoid any naming conflicts in the wild

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php]: Some notes:
* avoid the word "we" in the codebase
* no space between the last `@param` and `@return` for consistency within Core
* Add `"Optional"` to each optional's parameter description
* Add the default value in each optional `@param`
* `array<string>` is not a PHP type. Use `string[]` instead.

```suggestion
/**
 * Helper function to get the Template Hierarchy for the given slug.
 * Handles special cases like `front-page`, `singular` and `archive` templates.
 *
 * Note: Always add `index` as the last fallback template.
 *
 * @since 6.1.0
 *
 * @param string  $slug           The template slug to be created.
 * @param boolean $is_custom      Optional. Indicates if a template is custom or
 *                                part of the template hierarchy. Default false.
 * @param string $template_prefix The template prefix for the created template.
 *                                Used to extract the main template type, e.g.
 *                                in `taxonomy-books` the `taxonomy` is extracted.
 *                                Default empty string.
 * @return string[] The template hierarchy.
 */
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (hellofromtonya) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Avoid using the word "we" in the codebase.
```suggestion
							'description' => __( 'The template prefix for the created template. This is used to extract the main template type, e.g. in `taxonomy-books` extracts the `taxonomy`' ),
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (hellofromtonya) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: No empty space between param and return.
```suggestion
	/**
	 * Returns the fallback template for the given slug.
	 *
	 * @since 6.1.0
	 *
	 * @param WP_REST_Request $request The request instance.
	 * @return WP_REST_Response|WP_Error
	 */
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (hellofromtonya) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Remove `'gutenberg'` text domain.
```suggestion
				__( 'Invalid slug.' ),
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (hellofromtonya) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: These 2 conditionals can be combined as there's no other code within the first if block.

File: tests/phpunit/tests/block-template-utils.php
Comment (hellofromtonya) [inline tests/phpunit/tests/block-template-utils.php]: For each assertion in this test:

* Use `assertSame()` instead of `assertEquals`
* Add an assertion message (3rd arg) to describe it. Why? If an assertion fails, the rest will not run. By adding an assertion message, it helps to identify which assertion failed and why.

File: tests/phpunit/tests/rest-api/rest-post-types-controller.php
Comment (hellofromtonya) [inline tests/phpunit/tests/rest-api/rest-post-types-controller.php]: Add an assertion message, 3rd arg.

File: tests/phpunit/tests/rest-api/wpRestTemplatesController.php
Comment (hellofromtonya) [inline tests/phpunit/tests/rest-api/wpRestTemplatesController.php]: Add messages to each assertion.

File: tests/phpunit/tests/rest-api/wpRestTemplatesController.php
Comment (hellofromtonya) [inline tests/phpunit/tests/rest-api/wpRestTemplatesController.php]: * use `assertSame()` instead of `assertEquals()`
* add a message to each assertion

**Review by hellofromtonya — CHANGES_REQUESTED** [review]
First pass through the code. Most changes requested are for compliance to Core's coding / testing standards and/or for consistency.

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [reply src/wp-includes/block-template-utils.php:1047]: Oh I see, the other function names in that same file do not use the `wp_` prefix. Interesting. Okay, seems it's consistent within its context.

**Comment by hellofromtonya** [comment]
@spacedmonkey Are you happy with the REST code in this PR?

File: tests/phpunit/tests/block-template-utils.php
Comment (ntsekouras) [reply tests/phpunit/tests/block-template-utils.php]: I updated all the test suggestions. Thanks for explaining the `why` for multiple assertions.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php]:
```suggestion
		if ( rest_is_field_included( 'icon', $fields ) ) {
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:396]: Does this need to be in the view context? The icon is not public right now?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]:
```suggestion
		$response = $this->prepare_item_for_response( $fallback_template, $request );
		return rest_ensure_response( $response );
```

Please ensure that response is run through `prepare_item_for_response`, so the response is formatted correctly.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: This check should not be here. If you set `'required' => true,` to then this check should not be required. 

What if the slug was 0?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:80]:
```suggestion
							'description' => __( 'The slug of the template to get the fallback for' ),
							'required' => true,
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php]: This value can be string or null?

File: tests/phpunit/tests/block-template-utils.php
Comment (spacedmonkey) [inline tests/phpunit/tests/block-template-utils.php]: Add covers comments here.

File: tests/phpunit/tests/rest-api/wpRestTemplatesController.php
Comment (spacedmonkey) [inline tests/phpunit/tests/rest-api/wpRestTemplatesController.php:690]: Add covers.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php]: Should this be updated to all the other checks? If yes, it would be better to do in a separate PR, no?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Should this be `get_item_permissions_check`

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:396]: What do you mean public? I think it would be good to have in `view` as it can be useful in some cases. Do you see any drawback being in `view` context?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: I'll update but this change will also need a GB PR(therefore a packages update) to retrieve the content from `content.raw`.

File: tests/phpunit/tests/rest-api/wpRestTemplatesController.php
Comment (ntsekouras) [reply tests/phpunit/tests/rest-api/wpRestTemplatesController.php:690]: I'm not sure what more to add besides the existing one.. Any suggestions?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php]: It seems so if the post type hasn't declared a `menu_icon`. Should we update the `type` here? Do we need conditional adding to the response if the `menu_icon` is set? How do you handle this in general in core?

**Comment by ockham** [comment]
Thank you @spacedmonkey and @ntsekouras!

Unfortunately, it seems like [unit tests are failing](https://github.com/WordPress/wordpress-develop/actions/runs/3089920029/jobs/4998078712):

```
1) Tests_REST_WpRestTemplatesController::test_get_template_fallback
Trying to get property 'slug' of non-object

/var/www/tests/phpunit/tests/rest-api/wpRestTemplatesController.php:766
phpvfscomposer:///var/www/vendor/phpunit/phpunit/phpunit:97
/var/www/vendor/bin/phpunit:118
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php]: You can have multiple types. See 
https://github.com/WordPress/wordpress-develop/blob/61f7e758952dabbe7e00cd2d9eeca2e7d82e4604/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2218

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php]: I don't mind doing this in this PR.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:396]: If you make a property as view it means it can't be viewed pubilically. Do we want to expose this data publically?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:225]: If this value is null, should it maybe default to something?

Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:225]: I think it's okay to be null since it's not required and the consumers can decide what to do with it. For example in php core for post types we assign some default icon if it's in the `admin menu`, but in the site editor we assign different default icon..

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:396]: I see no harm in it, but if you have a strong opinion on this, I can remove it.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Make the change and backport it. I can approve the change there as well. Change the properties / shape of response can / will cause problems down the road.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:80]: There are now two required fields here.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: GB PR is here: https://github.com/WordPress/gutenberg/pull/44299 --cc @ockham

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php
Comment (ntsekouras) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:80]: Good catch, thanks!

**Review by spacedmonkey — APPROVED** [review]

**Comment by hellofromtonya** [comment]
Currently reviewing. When ready, will prep the commit.

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php]:
```suggestion
		list( $type ) = explode( '-', $template_prefix );
```

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php]: Remove "we"

```suggestion
		// These checks are needed because the `$slug` above is always added.
```

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php:1058]:
```suggestion

	$template_hierarchy = array( $slug );

```

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php:1072]:
```suggestion

	// Handle `archive` template.
```

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php:1096]:
```suggestion

	$template_hierarchy[] = 'index';

```

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php:1087]:
```suggestion

	// Handle `singular` template.
```

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php]:
```suggestion
 * @param string $template_prefix Optional. The template prefix for the created template.
```

**Comment by hellofromtonya** [comment]
Rebased the branch on top of the latest `trunk`.

**Review by hellofromtonya — APPROVED** [review]
LTGM 👍 Ready for commit, which I'm prepping next.

**Comment by hellofromtonya** [comment]
Note: The npm CI job failure is unrelated to this PR.

File: src/wp-includes/block-template-utils.php
Comment (hellofromtonya) [inline src/wp-includes/block-template-utils.php]:
```suggestion
 * Gets the template hierarchy for the given template slug to be created.
 *
```

**Comment by hellofromtonya** [comment]
Committed via changeset https://core.trac.wordpress.org/changeset/54269. Thank you everyone for your contributions!

**Pull request closed without merging**
